frame/support/tests/versioned_runtime_upgrade.rs

```diff
@@ -0,0 +1,208 @@
+};
+use frame_system::Config;
+use sp_core::{ConstU16, Get};
+
+// Because `derive_impl` is a [macro_magic](https://crates.io/crates/macro_magic) attribute
+// macro, [`#[use_attr]`](`frame_support::macro_magic::use_attr`) must be attached to any use
+// statement that brings it into scope.
+#[use_attr]
+use frame_support::derive_impl;
+
+type UncheckedExtrinsic = frame_system::mocking::MockUncheckedExtrinsic<Test>;
+type Block = frame_system::mocking::MockBlock<Test>;
+
+#[frame_support::pallet]
+mod dummy_pallet {
+	use frame_support::pallet_prelude::*;
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(4);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {}
+
+	#[pallet::storage]
+	pub type SomeStorage<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::genesis_config]
+	#[derive(Default)]
+	pub struct GenesisConfig {}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {}
+	}
+}
+
+impl dummy_pallet::Config for Test {}
+
+construct_runtime!(
+	pub enum Test where
+		Block = Block,
+		NodeBlock = Block,
+		UncheckedExtrinsic = UncheckedExtrinsic,
+	{
+		System: frame_system::{Pallet, Call, Config, Storage, Event<T>} = 0,
+		DummyPallet: dummy_pallet::{Pallet, Config, Storage} = 1,
+	}
+);
+
+#[derive_impl(frame_system::config_preludes::TestDefaultConfig as frame_system::DefaultConfig)]
+impl frame_system::Config for Test {
+	type BaseCallFilter = frame_support::traits::Everything;
+	type RuntimeOrigin = RuntimeOrigin;
+	type RuntimeCall = RuntimeCall;
+	type RuntimeEvent = RuntimeEvent;
+	type PalletInfo = PalletInfo;
+	type OnSetCode = ();
+}
+
+pub(crate) fn new_test_ext() -> sp_io::TestExternalities {
+	let storage = frame_system::GenesisConfig::default().build_storage::<Test>().unwrap();
+	let mut ext: sp_io::TestExternalities = sp_io::TestExternalities::from(storage);
+	ext.execute_with(|| System::set_block_number(1));
+	ext
+}
+
+/// A dummy migration for testing the `VersionedRuntimeUpgrade` trait.
+/// Sets SomeStorage to S.
+struct SomeUnversionedMigration<T: Config, S: Get<u32>>(sp_std::marker::PhantomData<(T, S)>);
+
+parameter_types! {
+	const UpgradeReads: u64 = 4;
+	const UpgradeWrites: u64 = 2;
+}
+
+/// Implement `OnRuntimeUpgrade` for `SomeUnversionedMigration`.
+/// It sets SomeStorage to S, and returns a weight derived from UpgradeReads and UpgradeWrites.
+impl<T: dummy_pallet::Config, S: Get<u32>> OnRuntimeUpgrade for SomeUnversionedMigration<T, S> {
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		dummy_pallet::SomeStorage::<T>::put(S::get());
+		RocksDbWeight::get().reads_writes(UpgradeReads::get(), UpgradeWrites::get())
+	}
+}
+
+type VersionedMigrationV0ToV1 = VersionedRuntimeUpgrade<
+	ConstU16<0>,
+	ConstU16<1>,
+	SomeUnversionedMigration<Test, ConstU32<1>>,
+	DummyPallet,
+	RocksDbWeight,
+>;
```

Comment:
I think this looks good. Associated consts would actually make it more complicated, or?  
But instead of `ConstU16` it should hopefully be possible to use `const u16` as Rust generic.

Reply:
`const u16` seems to work. The only potential issue is Rust wants me to uppercase them, so we'd have FROM and TO instead of From and To. WDYT?

Reply:
I think we only used one letter generics for types; `FROM` and `TO` is good.

Reply:
Refactored to use native rust generics for FROM and TO